taxjar/taxjar-magento2-extension
# INT 2085: Optimize transaction backfill order search

### Context

Current `BackfillTransactions:class` observer uses SearchCriteria to retrieve orders within given date and store constraints, but does not consider whether the order is actually "syncable" before batching that order into a bulk operation. The consumer `Backfill::class` then determines whether or not we should sync the transaction. This means that potentially many unnecessary operations are created only to determine that the orders within should not be synced.

### Description

This PR will now check within `BackfillTransactions:class` whether or not "force" option was enabled, and if force is not enabled, only syncable orders will be included in operations. While adding this functionality, `BackfillTransactions.php` was refactored, unit and integration tested.

_Also corrects failing Transaction::class unit tests_

### Performance

The same total number of order-date comparisons will now be paid upfront instead spread across consumer processes, but there should be a net decrease in number of operations for the same request.

### Testing

#### Automated Testing
-  Adds `Taxjar\SalesTax\Test\Unit\Observer\BackfillTransactionsTest.php`
-  Adds `Taxjar\SalesTax\Test\Integration\Observer\BackfillTransactionsTest.php`

#### Manual Testing
1. Enable Taxjar transaction sync
2. Create/invoice/ship an order (optionally create credit memo as well)
3. Observe order is synced with Taxjar
4. Trigger backfill transaction sync (_not forced_) with date range inclusive of order
5. Observe created order not included in any created operations
6. Trigger _forced_ backfill transaction sync with date range inclusive of order
7. Observe created order is now included in created operations

#### Versions
